rs/nns/governance/src/neuron_store.rs

```diff
@@ -555,37 +555,50 @@ impl NeuronStore {
         neuron_id: NeuronId,
         sections: NeuronSections,
     ) -> Result<(Cow<Neuron>, StorageLocation), NeuronStoreError> {
-        let heap_neuron = self.heap_neurons.get(&neuron_id.id).map(Cow::Borrowed);
-
-        if let Some(heap_neuron) = heap_neuron.clone() {
-            // If the neuron is active on heap, return early to avoid any operation on stable
-            // storage. The StableStorageNeuronValidator ensures that active neuron cannot also be
-            // on stable storage.
-            if !heap_neuron.is_inactive(self.now()) {
-                return Ok((heap_neuron, StorageLocation::Heap));
+        let main = || {
```

Comment:
Did anything change in the logic here?  I found the explicit match statement that was there easier to reason about

Reply:
+1. I find it unnecessary to have a reader reason about a stack.

Reply:
Yes. This fixes the bug mentioned in the description: before, there was a case where normalization was not applied. That happened if line 565 was executed. To address that, main makes sure that ALL return paths go through normalization.

One of the tangential changes is replacing match with result_copies. Let me set it back to match, since two people prefer match.

# Rationale for result_copies

I feel that

```
match (foo, bar) { ...
```

almost always inflates the number of "truly distinct" cases. It is super rare that the _combination_ of foo and bar is needed. They are almost always orthogonal. I think this is an example of that. E.g. it could be implemented like this:

```
if Some(stable) = stable {

    // From the caller's point of view, this has no effect.
    if Some(heap) = heap {
        log();
    }

    return stable;
}

if Some(heap) = heap {
    heap
} else {
    NotFound
}
```

This more clearly shows that they are orthogonal: To determine what to return, first, stable is considered, _then_ heap.

What we are really concerned about here is _how many_ copies there are. (1 is ideal. 2 is less than ideal, but we can work with it. 0 means that the caller is probably committing (the equivalent of) an HTTP 404. The way that we deal with 2 is to log, but otherwise pretend that we are in the 1 case.) In fact, from the caller's point of view, we really only distinguish between 0 and positive. My refactor attempted to reflect this.

Reply:
It seems we have different views of what is 'orthogonal'. In the example of the alternative with if-statements, I think it would only be orthogonal if there is no nested if's (i.e. not having `if Some(heap) = heap {` nested within `if Some(stable) = stable {`). Definitions aside, I think a more practical reason to use match is just to avoid nested if's which increase the cognitive load.